msprime/trees.py

```diff
@@ -159,6 +177,34 @@ def __str__(self):
 
 
 class EdgesetTable(_msprime.EdgesetTable):
+    """
+    Class for tables describing all edgesets in a tree sequence, of the form
+        left	right	parent	children
+        5.0	7.0	4	0,1
+        2.0	2.5	5	1,3
+        3.2	3.5	4	0,1
+        0.0	3.2	6	0,1,3
+    These describe the half-open genomic interval affected: `[left, right)`,
+    the `parent` and the `children` on that interval.
+
+    Requirements: to describe a valid tree sequence, a `EdgesetTable` (and
+    corresponding `NodeTable`, to provide birth times) must satisfy:
+        1. each list of children must be in sorted order,
+        2. any two edgesets that share a child must be nonoverlapping, and
+        3. the birth times of the `parent` in an edgeset must be strictly
+            greater than the birth times of the `children` in that edgeset.
+    Furthermore, for algorithmic requirements
+        4. the the table must be sorted by birth time of the `parent`, and
+        5. any two edgesets corresponding to the same `parent` must be nonoverlapping.
+    It is an additional requirement that the complete ancestry of each sample
+    must be specified, but this is harder to verify.
+
+    It is not required that all records corresponding to the same parent be
```

Comment:
> I am already advocating for having two sets of requirements, "strict" and "relaxed", with "relaxed" tree sequences not needing to satisfy the last requirement, but with simplify and simulate guaranteed to output strict tree sequences, but methods other than simplify not guaranteed to work on relaxed ones. So one option would be to add more requirements to strict ones - I think that edgesets are already left-to-right ordered in output, no?

I see the attraction of having strict and relaxed tree sequences, but I worry about the consequences. Not guaranteed to work is easy to say, but I don't think that's enough. We would have to say, "guaranteed to fail predictably with a meaningful error". Just saying the behaviour is undefined definitely isn't good enough, and if we write efficient algorithms in C, then we need to be able to lean heavily on assumptions about the input. If we have to test that these assumptions are true before we run the algorithm, then this could be a heavy cost that we end up paying over and over again. Also, for every algorithm, we'd have to think, what are the possible input variations here, and how would they possibly end up leading to (e.g.) memory violations and so on. It just seems error prone to me to allow for needless variation.

I also like the idea of canonicalisation. When you have a complex combinatorial object, then I think it makes sense to remove as much symmetry as possible from it. Pick a single representation that has useful properties, and stick with it. This allows for meaningful definitions of equality, as well as the algorithmic considerations above.

> So, how about this: there will need to be a method that orders edgesets appropriately: by parent birth time at least; possibly by left endpoint as well. The only question is whether that is the user's responsibility or is the first step in simplify(). Either is fine, although it seems nicer and cleaner to put it in simplify, as otherwise we'll have to check if its true anyhow? What do you think?

Well, if we insist that all edgesets for a given parent are adjacent and they are ordered left-to-right within a parent, and edgesets parents are ordered by non-decreasing time (which we can check cheaply on import) then this first part of simplify is done. It's just:
```python
edgesets = ts.edgesets()
parent_edgesets = [next(edgesets)]
last_parent = parent_edgesets[0].parent
for edgeset in edgesets:
    if edgeset.parent != last_parent:
          yield parent_edgesets
          parent_edgesets = []
          last_parent = edgeset.parent
     parent_edgesets.append(edgeset)
yield parent_edgesets
```
The ``parent_edgesets`` list can be transformed directly into a segment chain, since the edgesets are guaranteed to be in left-to-right order. Therefore, we can implement simplify with no sorting **at all** (well, except for sorting children within an edgeset, but that's very cheap).

This seems to me like a very useful property to be able to provide, and I'm sure this will be useful in other algorithms besides simplify. Plus, we could also add a ``parent`` argument to the edgesets iterator which returns all edgesets for a given parent. This could either be implemented by skimming forward until we reach the parent in question, or by keeping an array of offsets to the start of the edgesets for each parent (a modest extra memory cost).

I know it's awkward for users doing the required sorting to get the input in the required form, but perhaps we could provide method to do this? I.e., given node, edgesets, site and mutation tables, if we run ``canonicalise`` on them, they do all the required sorting. That way, if you know that your tables have the required form then you can avoid this cost. If you are doing something awkward then you don't have to worry about what the precise requirements are and just fill your tables as suits your application. Running ``canonicalise`` will sort these tables in the correct orders, and they can then be imported into a valid tree  sequence.